template-files/action.py

```diff
@@ -270,32 +357,77 @@ def template_file(
             dst.parent.mkdir(parents=True, exist_ok=True)
             dst.write_text(template.render(**{**context, **standard_context}))
         except TemplateError as err:
-            perror(f"* ❌ Failed to template `{src}`: {err}", indent=INDENT)
+            perror(f"* :cross_mark: Failed to template `{src}`: {err}", indent=INDENT)
             return 1
 
-        print(f"* ✅ `{src}` → `{dst}`", indent=INDENT)
-
         # display stubs & context for this file
-        if stubs or context:
-            table = Table.grid(padding=1)
-            table.add_column()
-            table.add_column(max_width=TemplateState.WIDTH)
-            table.add_column()
-            for stub, state in stubs.items():
+        table = None
+        error = False
+        warning = False
+        if stubs or variables:
+            table = Table.grid(padding=(0, 1))
+            # stubs
+            for stub, count in stubs.items():
+                state = TemplateState.from_count(count)
                 table.add_row("*", state, f"`{stub}`")
-            for key, value in context.items():
-                table.add_row("*", TemplateState.CONTEXT, f"`{key}={value}`")
+            # variables
+            for variable, value in variables.items():
+                if variable in standard_context:
+                    continue
+                state = TemplateState.from_value(value)
+                if isinstance(value, Undefined):
+                    error = True
+                    value = ""
+                elif value is missing:
+                    warning = True
+                    value = ""
+                table.add_row("*", state, f"`{variable}={value}`")
+            for variable in set(context) - set(variables):
+                value = context[variable]
+                table.add_row("*", TemplateState.UNUSED, f"`{variable}={value}`")
+
+        if error:
+            perror(f"* :cross_mark: Context missing `{src}` → `{dst}`", indent=INDENT)
+        elif warning:
+            print(f"* :warning-emoji: `{src}` → `{dst}`", indent=INDENT)
+        else:
+            print(f"* :white_check_mark: `{src}` → `{dst}`", indent=INDENT)
+        if table:
             print(table, indent=INDENT * 2)
-    return 0  # no errors
+
+        return int(error)
+
+
+class LocalContents:
+    # mirror GitHub contents object
+    def __init__(self, path: Path):
+        self.decoded_content = path.read_text().encode()
+
+
+class LocalRepository:
+    # mirror GitHub repository object
+    def __init__(self, *paths: str | os.PathLike[str] | Path) -> None:
+        self.path = Path(*paths).resolve()
+        if not paths or not self.path.is_dir():
+            raise FileNotFoundError(f"{self.path} is not a directory")
+
+        # constants
+        self.html_url = f"file://{self.path}"
+        self.user = "<local>"
+        self.name = "<local>"
+
+    def get_contents(self, path: str) -> LocalContents:
+        return LocalContents(self.path / path)
```

Comment:
these mirror the GitHub Repository and Contents objects allowing us to use templates from the local file system (this is useful for testing purposes)